Orange/widgets/visualize/owmosaic.py

```diff
@@ -438,34 +439,29 @@ def set_data(self, data):
             self.init_combos(None)
             return
 
-        self.color_model.set_domain(self.data.domain)
         self.init_combos(self.data)
 
         self.openContext(self.data)
 
-        # if we first received subset we now call setSubsetData to process it
-        if self.unprocessed_subset_data:
-            self.set_subset_data(self.unprocessed_subset_data)
-            self.unprocessed_subset_data = None
-
-        self.set_color_data()
-
     @Inputs.data_subset
     def set_subset_data(self, data):
-        self.Warning.incompatible_subset.clear()
-        if self.data is None:
-            self.unprocessed_subset_data = data
-            return
-        try:
-            self.subset_data = data.transform(self.data.domain)
-        except:
-            self.subset_data = None
-            self.Warning.incompatible_subset(shown=data is not None)
+        self.subset_data = data
 
     # this is called by widget after setData and setSubsetData are called.
     # this way the graph is updated only once
     def handleNewSignals(self):
-        self.reset_graph()
+        self.Warning.incompatible_subset.clear()
+        self.subset_indices = None
+        if self.data is not None and self.subset_data:
+            try:
+                indices = {e.id for e in self.subset_data.transform(self.data.domain)}
+            except:
+                self.Warning.incompatible_subset(shown=self.subset_indices is not None)
```

Comment:
Have I overlooked something, or is `self.subset_indices` always `None` here? Originally, this was `data is not None`, but here even this is not needed since data is checked to be non-None.

Second, where can this `try` clause fail at all?

Perhaps add a test that feeds iris as data and titanic as subset, and check that the warning is displayed. (Currently it's not - I tried.)

A general problem: how do we defined an "incompatible subset"? I guess a subset is incompatible if none of the attributes can be transformed, that is, if `transformed = self.subset_data.transform(self.data.domain)` and `np.all(np.isnan(transformed.X)) and np.all(np.isnan(transformed.Y))`.